### ISSUE TITLE
Severe lag/performance issues for filetype .js, issues DO NOT persist when filetype converted to .txt

### ISSUE THREAD

**Issue by SKGrimes** [atom/atom#9753]
## Issue

I suspect there is a memory leak, or an otherwise non-trivial performance error occurring when I try to open [this javascript file](https://github.com/SKGrimes/heapshots/blob/master/sampledata.js). This error seems localized and specific to `filetype .js` as simply changing the file extension to `.txt` removes the error. The file will load instantly in `.txt` form but often takes 30+ seconds otherwise. If you access the file `sampledata.js.txt`, [this file](https://github.com/SKGrimes/heapshots/blob/master/sampledata.js.txt), after changing it to a .txt file, it loads immediately. Attempting to open the file in Atom takes **upwards of 30 seconds** and also has **crashed the editor** on severral occasions. It will render the following error every time:

<img width="874" alt="screen shot 2015-11-22 at 6 04 16 pm" src="https://cloud.githubusercontent.com/assets/7623078/11327084/75375ea6-9149-11e5-94e9-28d4de79b69d.png">
*note: this picture is taken of the error in safe mode, i have taste and therefore my theme is much slicker than this.*
There did not seem to be anything useful in the console, but I neglected to save it. From memory it said something like, amount of time to load the window, and ::shadow deprecated. However, the performance data has been saved in a repo. Here is all of the information to produce the error:

Atom Editor Version: **1.2.3 (latest)**
Problem Persists in Safe Mode: **true**
APM Links: **false; (no links)**

Time to Open File (stopwatch)
    attempt 1  : **55.02 seconds**
    attempt 2  : **36.43 seconds**
    attempt 3  : **35.88 seconds**
    average(mean): **42.44 seconds**

Settings 
   linter:              **disabled**
   linter-jshint:     **disabled**
   comments: **Behaviour persists in safe mode; appears to be default behaviour.**

### Performance Data
The performance Data has been saved in a repo for documentation purposes. Please find relevant data [here](https://github.com/atom/atom/files/41206/cpu_profile_started_pre_fileload_v003.cpuprofile.txt).

cpu profiles
* 2 profiles done while the file was actively loading
* 1 profile was started **before** trying to access the file.

heap snapshots
* pre-open means the snapshot was taken/started **before** clicking on the file/trying to access it in atom.
* after-open means the snapshot was taken/started **after** clicking on the file/trying to access it in atom.
https://github.com/SKGrimes/heapshots

[cpu_profile_actively_loading_v002.cpuprofile.txt](https://github.com/atom/atom/files/41207/cpu_profile_actively_loading_v002.cpuprofile.txt)
[cpu_profile_started_pre_fileload_v003.cpuprofile.txt]

**Also included is a heaptimeline.**

### FILE 
note that the file is in `.txt` form, where performance is not an issue. To reproduce error, the file should have the `.txt` extension removed. Then, when attempting to open it, the observed behaviour should occur on your machine.
[sampledata.js.txt](https://github.com/atom/atom/files/41208/sampledata.js.txt)

If the sample link fails to load you can retrieve a copy of the file from the repo  [here](https://github.com/atom/atom/files/41206/cpu_profile_started_pre_fileload_v003.cpuprofile.txt).

**Comment by 50Wliu** [comment]
Awesome bug report @SKGrimes!   Could you please re-upload the second profile (started_pre_fileload), since it didn't appear to complete successfully?  Thanks!

Also, would it be possible to see if you can isolate your issue to the smallest-possible JS file that causes the performance issues?

I'm also getting freezes while loading the file on Windows 10 (not safe mode).  I suspect it has something to do with the JSON-like code.

**Comment by SKGrimes** [comment]
@50Wliu 
Thanks for the speedy response mate. I just uploaded a set of manipulated sample data to the repo, you can find it within [this directory](https://github.com/SKGrimes/heapshots/tree/master/atom_bug_sample_data).

It isn't all super relevent, but basically within the [original file](https://github.com/SKGrimes/heapshots/blob/master/atom_bug_sample_data/full_sample.js) are two keys which are absolutely loaded:

```
"companyInfo": {
  "legalTerms" : ~52,000 chars long...
  "base64Logo": 19,000 chars long...
```

You will notice that if you try and open a file without the base64 logo; this [file](https://github.com/SKGrimes/heapshots/blob/master/atom_bug_sample_data/no_base64.js) there is still lag, but it is much improved. A decent control would be the file [no_big_keys](https://github.com/SKGrimes/heapshots/blob/master/atom_bug_sample_data/no_big_keys.js) which removes both the `base64Logo` key AND the `legalTerms` key. Performance is noticeably better.

As the file `base64Logo` key is the largest, if you open the file without the `legalTerms`, [this one](https://github.com/SKGrimes/heapshots/blob/master/atom_bug_sample_data/no_legal.js) performance is still pretty grim.

## tl;dr

Obviously, dumping JSON data into your text editor to see how long it hangs isn't super interesting. So the best I can really provide for you (so you may be spared the trouble of having to do it), is simply downloading [this file](https://github.com/SKGrimes/heapshots/blob/master/atom_bug_sample_data/max_key_char_limit.js) which is (supported only by anecdotal evidence here) around the peak before performance goes really wonky.

**The max char size for a key is around ~8K.** Performance is still lack luster, but loading and editing seem to limit lag to <1 second.

**Comments**: I don't understand why the size of the key matters here. This is a text editor but it does seem to realize that it is viewing JS. Even if I turn off `git-diff`, `linter`, `linter-jshint`, it still lags horribly on the main file. Somehow it is hating keys over ~8K chars **AND** is aware of them. i.e not treating text as dumb text.

**Comment by 50Wliu** [comment]
Aha, didn't even notice those two.  In that case, this would appear to be a duplicate of both #1667 and #979.
